httpbp/middlewares.go

```diff
@@ -299,3 +306,101 @@ func recoverPanic(name string, next HandlerFunc) HandlerFunc {
 		return next(ctx, w, r)
 	}
 }
+
+// statusCodeRecorder is used by RecordStatusCode to record the code passed to
+// a call to WriteHeader.
+type statusCodeRecorder struct {
+	http.ResponseWriter
+
+	code int
+}
+
+func (r *statusCodeRecorder) WriteHeader(code int) {
+	r.ResponseWriter.WriteHeader(code)
+	if r.code == 0 {
+		r.code = code
+	}
+}
+
+func (r *statusCodeRecorder) getCode(err error) int {
+	if r.code != 0 {
+		// WriteHeader was called explicitly, use that
+		return r.code
+	}
+	if err != nil {
+		// something went wrong, check if err is an HTTPErr where we can extract
+		// the code, otherwise assume InternalServerError
+		var httpErr HTTPError
+		if errors.As(err, &httpErr) {
+			return httpErr.Response().Code
+		}
+		return http.StatusInternalServerError
+
+	}
+	// if there's no error returned and no call to WriteHeader, Go will
+	// return OK.
+	// https://pkg.go.dev/net/http#ResponseWriter.WriteHeader
+	return http.StatusOK
+}
+
+var families = [6]string{
```

Comment:
🔕 is it really necessary to make this an array instead of slice? e.g. this should give you the same result?
```
var families = []string{
```
(unless there's performance implications)

Reply:
I don't know that it's necessary, it was just what @kylelemons example was like:

https://github.com/reddit/baseplate.go/pull/453#discussion_r771585560

Reply:
Use `[...]` instead of `[6]`.  Using the array turns `len` into a constant value and can improve inlining and bounds check elision.

Reply:
> Use [...] instead of [6].

:til:

Reply:
It actually does seem to be technically faster (albeit the numbers are very small):

slice:

```
goos: darwin
goarch: amd64
pkg: github.com/reddit/baseplate.go/httpbp
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkGetStatusCodeFamily
BenchmarkGetStatusCodeFamily-16    	858927290	         1.402 ns/op
PASS
```

array:
```
goos: darwin
goarch: amd64
pkg: github.com/reddit/baseplate.go/httpbp
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkGetStatusCodeFamily
BenchmarkGetStatusCodeFamily-16    	1000000000	         0.2339 ns/op
PASS
```